### PR TITLE
[promptflow][internal] Fix standard OpenTelemetry traces collection and line run list bug

### PR DESCRIPTION
**Pull request by zhengfeiwang**
# Description

**Bug fix**

Make sure line runs covering all traces.

**Compatible with standard trace**

Use `get()` to avoid `KeyError` when handle standard trace without `session_id` and `span_type` in attributes.

**Copy button**

Add a copy button at the very top of the trace UI page, so that we can easily copy the JSON to UX demo page for rendering.

image

**Order by start time**

List traces order by `start_time`.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow/promptflow/_sdk/entities/_trace.py [hunk 1]
```diff
@@ -201,14 +202,15 @@ def _from_root_span(span: Span) -> "_LineRunData":
...
             trace_id=span.trace_id,
             root_span_id=span.span_id,
-            inputs=json.loads(attributes[SpanAttributeFieldName.INPUTS]),
-            outputs=json.loads(attributes[SpanAttributeFieldName.OUTPUT]),
+            # for standard OpenTelemetry traces, there won't be `inputs` and `outputs` in attributes
+            inputs=json.loads(attributes.get(SpanAttributeFieldName.INPUTS, dict())),
+            outputs=json.loads(attributes.get(SpanAttributeFieldName.OUTPUT, dict())),
             start_time=start_time,
             end_time=end_time,
             status=span._content[SpanFieldName.STATUS][SpanStatusFieldName.STATUS_CODE],
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_trace.py:210]: shall this also using get?

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/entities/_trace.py:210]: no, this is from db, and they are consistent in our db

File: src/promptflow/promptflow/_sdk/entities/_trace.py [hunk 2]
```diff
@@ -124,6 +131,16 @@ def _from_protobuf_object(obj: PBSpan, resource: typing.Dict) -> "Span":
             SpanStatusFieldName.STATUS_CODE: parse_otel_span_status_code(obj.status.code),
         }
         attributes = flatten_pb_attributes(span_dict[SpanFieldName.ATTRIBUTES])
+        if SpanAttributeFieldName.SESSION_ID not in attributes:
```
Comment (wangchao1230) [inline src/promptflow/promptflow/_sdk/entities/_trace.py:134]: why not write this in two if else. Current logic assumes session_id & span_type should coexist.

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/entities/_trace.py:134]: good catch, let me refine this

File: src/promptflow/promptflow/_sdk/operations/_trace_operations.py [hunk 3]
```diff
@@ -13,18 +16,42 @@ class TraceOperations:
...
+                # standard OpenTelemetry trace, regard as a line run
+                grouped_orm_spans[first_orm_span.trace_id] = copy.deepcopy(orm_spans)
+            elif SpanAttributeFieldName.LINE_RUN_ID in attributes:
+                # for other traces, group by `line_run_id` as a line run
+                line_run_id = attributes[SpanAttributeFieldName.LINE_RUN_ID]
+                if line_run_id not in grouped_orm_spans:
+                    grouped_orm_spans[line_run_id] = []
+                grouped_orm_spans[line_run_id].extend(copy.deepcopy(orm_spans))
+            else:
+                referenced_line_run_id = attributes[SpanAttributeFieldName.REFERENCED_LINE_RUN_ID]
```
Comment (brynn-code) [inline src/promptflow/promptflow/_sdk/operations/_trace_operations.py:50]: We may skip this branch.

Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/operations/_trace_operations.py:57]: good catch, updated the condition and skip some cases.

File: src/promptflow/promptflow/_sdk/entities/_trace.py [hunk 4]
```diff
@@ -216,48 +230,35 @@ class LineRun:
     evaluations: typing.Optional[typing.List[typing.Dict]] = None
 
     @staticmethod
-    def _from_spans(spans: typing.List[Span]) -> typing.List["LineRun"]:
-        line_run_data_list: typing.List[_LineRunData] = []
-        evaluation_line_run_data_dict = dict()  # line run id -> {evaluation name -> eval line run data}
+    def _from_spans(spans: typing.List[Span]) -> "LineRun":
+        main_line_run_data: _LineRunData = None
+        evaluation_line_run_datas = dict()
```
[the remark this answers is not in the document]
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/entities/_trace.py:235]: good catch!

[the remark this answers is not in the document]
Comment (zhengfeiwang) [reply src/promptflow/promptflow/_sdk/entities/_trace.py:235]: fixed with `...data_dict` here

**Review by wangchao1230 — APPROVED** [review]

**Review by brynn-code — APPROVED** [review]

**Pull request merged**